### PR TITLE
Add custom nft bridge

### PR DESCRIPTION
**Pull request by trangtran-enya**
:clipboard: Add associated issues, tickets, docs URL here.

## Overview

One of our partner requests an example of NFT bridge for a custome ERC721
## Changes

Add a new example to boba-examples

## Testing

Describe how to test your new feature/bug fix and if possible, a step by step guide on how to demo this.

### CODE REVIEW

File: boba_examples/boba-custom-nft-bridge/contracts/L1NFTBridge.sol [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+        // if (pairNFTInfo[_l1Contract].baseNetwork == Network.L1) {
+        //     // check the existence of bridgeExtraData(uint256) on l1Contract
+        //     if (ERC165Checker.supportsInterface(_l1Contract, 0x9b9284f9)) {
+        //         extraData = iSupportBridgeExtraData(_l1Contract).bridgeExtraData(_tokenId);
+        //     } else {
+        //         // otherwise send tokenURI return (encoded in bytes)
+        //         // allow to fail if the call fails
+        //         extraData = abi.encode(IERC721Metadata(_l1Contract).tokenURI(_tokenId));
+        //     }
+        // }
```
Comment (trangtran-enya) [inline boba_examples/boba-custom-nft-bridge/contracts/L1NFTBridge.sol]: Looks like it is not used. So I plan to remove this? @Boyuan-Chen can u please confirm if it's correct.

File: boba_examples/boba-custom-nft-bridge/contracts/L2NFTBridge.sol [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+        // if (pairNFTInfo[_l2Contract].baseNetwork == Network.L2) {
+        //     // check the existence of bridgeExtraData(uint256) on l2Contract
+        //     if (ERC165Checker.supportsInterface(_l2Contract, 0x9b9284f9)) {
+        //         extraData = iSupportBridgeExtraData(_l2Contract).bridgeExtraData(_tokenId);
+        //     } else {
+        //         // otherwise send tokenURI return (encoded in bytes)
+        //         // allow to fail if the call fails
+        //         extraData = abi.encode(IERC721Metadata(_l2Contract).tokenURI(_tokenId));
+        //     }
+        // }
```
Comment (trangtran-enya) [inline boba_examples/boba-custom-nft-bridge/contracts/L2NFTBridge.sol]: Looks like its not used so I plan to remove.

File: boba_examples/boba-custom-nft-bridge/contracts/L2NFTBridge.sol [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+        // size limits unchecked
```
Comment (trangtran-enya) [inline boba_examples/boba-custom-nft-bridge/contracts/L2NFTBridge.sol]: same as above

File: boba_examples/boba-custom-nft-bridge/contracts/L1NFTBridge.sol (hunk 1 above)
Comment (boyuan-chen) [reply boba_examples/boba-custom-nft-bridge/contracts/L1NFTBridge.sol]: Yup! @souradeep-das Please confirm!

File: boba_examples/boba-custom-nft-bridge/contracts/CustomERC721.sol [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    mapping(uint256 => string) public feature_1;
+    mapping(uint256 => string) public feature_2;
+    mapping(uint256 => string) public feature_3;
+
+    function mint(address _to, uint256 _tokenId, bytes memory _data) external {
+        _safeMint(_to, _tokenId);
+
+        feature_1[_tokenId] = "1";
+        feature_2[_tokenId] = "2";
+        feature_3[_tokenId] = "3";
```
Comment (trangtran-enya) [inline boba_examples/boba-custom-nft-bridge/contracts/CustomERC721.sol:27]: Why do we fixate these values on L1? 🤔

**Comment by souradeep-das** [comment]
looking at the custom ERC721 here, and the features_1,2,3 requirements, looks like the current bridge could support this functionality? The respective bridgeWithExtraData() method on the current bridge would allow to encode and move any kind of custom data along with the token, and use the data however they decide (although, it needs them to implement iSupportBridgeExtraData) - is it that they have already deployed? in which case please disregard my comments
https://github.com/bobanetwork/boba/tree/develop/boba_examples/nft_bridging#making-your-erc721-bridge-extra-data

File: boba_examples/boba-custom-nft-bridge/contracts/L1NFTBridge.sol (hunk 1 above)
Comment (souradeep-das) [reply boba_examples/boba-custom-nft-bridge/contracts/L1NFTBridge.sol]: yeah makes sense to remove them if we don't use them
but regarding the usage - I was thinking that the customData could enforce their custom NFTs instead? unless they have already deployed https://github.com/bobanetwork/boba/pull/332#issuecomment-1235825009

Comment (boyuan-chen) [reply boba_examples/boba-custom-nft-bridge/contracts/L1NFTBridge.sol]: They have already deployed on Avalanche.

File: boba_examples/boba-custom-nft-bridge/contracts/CustomERC721.sol (hunk 4 above)
Comment (boyuan-chen) [reply boba_examples/boba-custom-nft-bridge/contracts/CustomERC721.sol:27]: It’s an example. They said they have some skill values that are different from the standard ERC721.

File: boba_examples/boba-custom-nft-bridge/contracts/L1NFTBridge.sol (hunk 1 above)
Comment (souradeep-das) [reply boba_examples/boba-custom-nft-bridge/contracts/L1NFTBridge.sol]: ohh got it, then this is the only way. perfect, all good to remove the extraData part from the example then! 🚀

File: boba_examples/boba-custom-nft-bridge/contracts/CustomERC721.sol (hunk 4 above)
Comment (trangtran-enya) [reply boba_examples/boba-custom-nft-bridge/contracts/CustomERC721.sol:27]: I mean this in constructor, we disallow creating customized features but always fixate on some values. If it's due to the ERC721 standard, I just wonder how other people handle if they want to customize these values.

**Comment by trangtran-enya** [comment]
> looking at the custom ERC721 here, and the features_1,2,3 requirements, looks like the current bridge could support this functionality? The respective bridgeWithExtraData() method on the current bridge would allow to encode and move any kind of custom data along with the token, and use the data however they decide (although, it needs them to implement iSupportBridgeExtraData) - is it that they have already deployed? in which case please disregard my comments https://github.com/bobanetwork/boba/tree/develop/boba_examples/nft_bridging#making-your-erc721-bridge-extra-data

I am not aware of  their needs in this case. Maybe we should ask them 🤔 ?

File: boba_examples/boba-custom-nft-bridge/contracts/CustomERC721.sol (hunk 4 above)
Comment (boyuan-chen) [reply boba_examples/boba-custom-nft-bridge/contracts/CustomERC721.sol:27]: Could you please remove the unnecessary code? I will ask them!

File: boba_examples/boba-custom-nft-bridge/contracts/L1NFTBridge.sol (hunk 1 above)
Comment (trangtran-enya) [reply boba_examples/boba-custom-nft-bridge/contracts/L1NFTBridge.sol]: Remove extraData

File: boba_examples/boba-custom-nft-bridge/contracts/CustomERC721.sol (hunk 4 above)
Comment (trangtran-enya) [reply boba_examples/boba-custom-nft-bridge/contracts/CustomERC721.sol:27]: @Boyuan-Chen I have removed.

**Comment by boyuan-chen** [comment]
They can upgrade their L1 NFTs to build `customData` payload and send to L2.

Closed.

**Pull request closed without merging**